rosetta-source/src/main/rosetta/event-common-func.rosetta

```diff
@@ -1661,6 +1661,173 @@ func Qualify_Shaping: <"The qualification of a shaping event from the fact that
+	inputs:
+		tradeState TradeState (1..1) <"The original trade to be repriced.">
+		newAllinPrice number(1..1) <"The collateral all-in price.">
+		newAssetQuantity number(1..1) <"The collateral quantity.">
+		newCashValue number(1..1)<"The new collateral value.">
+		effectiveRepriceDate AdjustableOrRelativeDate (1..1) <"The date to reprice the collateral">
+	output:
+		instruction PrimitiveInstruction (1..1)
+
+	alias oldPriceQuantity:
+		tradeState -> trade -> tradableProduct -> tradeLot-> priceQuantity
+		
+	alias currentPrice: <"Filter interest rate price and make it into a single element">
+		oldPriceQuantity
+		extract [ item -> price ]
+		flatten
+		filter [ item -> priceExpression -> priceType = PriceTypeEnum -> InterestRate ] only-element
+		
+	alias newPrice: <"Create the new price object.">
+		Create_Price( newAllinPrice, currentPrice -> unit, currentPrice -> perUnitOf, currentPrice -> priceExpression )
+		
+	alias changeQuantity:	<"Create distinct list of Quantity with value set to newAssetQuantity">
+		if newAssetQuantity <> 0 then
+			tradeState -> trade -> tradableProduct -> tradeLot only-element -> priceQuantity -> quantity
+				map [ Create_NonNegativeQuantitySchedule( newAssetQuantity, item -> unit ) ]
+				distinct
+		else 
+			empty
+			
+	alias newPriceQuantity: <"Create the new price quantity object.">
+		Create_PriceQuantity( [ newPrice ], changeQuantity, empty )
+		
+	condition ContractualProduct: <"This repricing function applies only to contractual products.">
+		tradeState -> trade -> tradableProduct -> product -> contractualProduct exists
+		
+	condition SingleTradeLot: <"This repricing function applies only to trades with a single lot.">
+		tradeState -> trade -> tradableProduct -> tradeLot count = 1
+	
+	
+	//Based on the ICMA repo lifecycle event processing, if the cash amount is being changed then the contract will be closed and a new one opened.
+	//If the collateral qualtity is changed and the cash amount is not then the contract is ammended. 
+	//In either cash the all-in price on the collateral may or may not be changed.
+	
+		
+	//Create new instruction to update the collateral price only of the cash amount does not change
+	set instruction -> quantityChange: <"Set primitive instruction to be only a price change.">
+		if newCashValue = 0 and newAssetQuantity = 0 then
+			Create_QuantityChangeInstruction( [newPriceQuantity], QuantityChangeDirectionEnum->Replace, empty )
+	
+	//Create new instruction if the cash does not change but the collateral quantity changes.
+	set instruction -> quantityChange: <"Set primitive instruction to be only a quantity change instruction.">
+		if newCashValue = 0 and newAssetQuantity <> 0 then
+			Create_QuantityChangeInstruction( [newPriceQuantity], QuantityChangeDirectionEnum->Replace, empty )
+			
+	//If the cash amount is changed then contract is closed and new one at the new cash amount is opened.
+	
```

Comment:
we can delete these two empty lines I think, as the comment above here refers to the next "set instruction:" clause